lib/FrontendTool/FrontendTool.cpp

```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+// 1. Paths to swiftmodules in the module cache or in the prebuilt cache are not
+//    recorded - Precondition: the corresponding path to the swiftinterface must
+//    already be present as a key in pathToModuleDecl.
+// 2. swiftmodules next to a swiftinterface are saved if they are up-to-date.
+void computeSwiftModuleTraceInfo(
+    const SmallPtrSetImpl<ModuleDecl *> &importedModules,
+    const llvm::DenseMap<StringRef, ModuleDecl *> &pathToModuleDecl,
+    const DependencyTracker &depTracker,
+    StringRef prebuiltCachePath,
+    std::vector<SwiftModuleTraceInfo> &traceInfo) {
+
+  SmallString<256> buffer;
+
+  std::string errMsg;
+  llvm::raw_string_ostream err(errMsg);
+
+  auto errorUnexpectedPath = [&err, &pathToModuleDecl]() {
+    err << "The module <-> path mapping we have is:\n";
+    for (auto &m: pathToModuleDecl)
+      err << m.second->getName() << " <-> " << m.first << '\n';
+    llvm::report_fatal_error(err.str());
+  };
+
+  using namespace llvm::sys;
+
+  auto computeAdjacentInterfacePath = [](SmallVectorImpl<char> &modPath) {
+    auto swiftInterfaceExt =
+      file_types::getExtension(file_types::TY_SwiftModuleInterfaceFile);
+    path::replace_extension(modPath, swiftInterfaceExt);
+  };
+
+  for (auto &depPath : depTracker.getDependencies()) {
+
+    // Decide if this is a swiftmodule based on the extension of the raw
+    // dependency path, as the true file may have a different one.
+    // For example, this might happen when the canonicalized path points to
+    // a Content Addressed Storage (CAS) location.
+    auto moduleFileType =
+      file_types::lookupTypeForExtension(path::extension(depPath));
+    auto isSwiftmodule =
+      moduleFileType == file_types::TY_SwiftModuleFile;
+    auto isSwiftinterface =
+      moduleFileType == file_types::TY_SwiftModuleInterfaceFile;
+
+    if (!(isSwiftmodule || isSwiftinterface))
+      continue;
+
+    auto dep = pathToModuleDecl.find(depPath);
+    if (dep != pathToModuleDecl.end()) {
+      // Great, we recognize the path! Check if the file is still around.
+
+      ModuleDecl *depMod = dep->second;
+      if(depMod->isResilient() && !isSwiftinterface) {
+        SmallString<256> moduleAdjacentInterfacePath(depPath);
+        computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+        if (fs::real_path(moduleAdjacentInterfacePath, buffer, true)) {
+          err << "The module " << depMod->getName().str() << " has library"
+              << " evolution enabled but we're recording a non-adjacent"
+              << " swiftmodule at\n" << depPath << "\nin the trace.";
+          llvm::report_fatal_error(err.str());
+        }
+        buffer.clear();
+      }
+
+      // FIXME: Better error handling
+      StringRef realDepPath
+        = fs::real_path(depPath, buffer, /*expand_tile*/true)
+        ? StringRef(depPath) // Couldn't find the canonical path, assume
+                             // this is good enough.
+        : buffer.str();
+      buffer.clear();
```

Comment:
You can't clear the buffer; it's backing the StringRef.

Reply:
Wow, I'm surprised to see that StringRef doesn't copy the data when you give it a raw pointer... How was this code even working before then? https://github.com/apple/swift/blob/a51162290e9df4507c1d5dcce58a0958dac4d785/lib/FrontendTool/FrontendTool.cpp#L284-L311

Reply:
StringRef is string_view, the point is that it does not copy the data. I assume the old code worked because `real_path` is clearing the buffer contents itself, but it certainly doesn't say that in its doc comment. :-/

Reply:
If `real_path` were clearing the buffer, then we would've printed out the same path for every module, which wasn't happening earlier, because all the `realDepPath`s saved in the `swiftModules` vector point to the beginning of `buffer`.

Reply:
SwiftModuleTraceInfo stores a std::string for the path; that's where it gets copied out.